authzen/client.go

```diff
@@ -0,0 +1,439 @@
+type Opt func(*Client)
+
+// WithHTTPClient sets a custom HTTP client.
+func WithHTTPClient(client *http.Client) Opt {
+	return func(c *Client) {
+		c.httpClient = client
+	}
+}
+
+// WithTimeout sets the request timeout.
+func WithTimeout(timeout time.Duration) Opt {
+	return func(c *Client) {
+		c.httpClient.Timeout = timeout
+	}
+}
+
+// WithHeaders sets custom headers to be included in all requests.
+func WithHeaders(headers map[string]string) Opt {
+	return func(c *Client) {
+		maps.Copy(c.headers, headers)
+	}
+}
+
+// WithUserAgent sets a custom user agent string.
+func WithUserAgent(userAgent string) Opt {
+	return func(c *Client) {
+		c.userAgent = userAgent
+	}
+}
+
+func WithTLSConfig(tlsConf *tls.Config) Opt {
+	return func(c *Client) {
+		if t, _ := c.httpClient.Transport.(*http.Transport); t == nil {
+			c.httpClient.Transport = &http.Transport{
+				TLSClientConfig: tlsConf,
+			}
+		} else {
+			t.TLSClientConfig = tlsConf
+		}
+	}
+}
+
+// WithTLSInsecure enables skipping TLS certificate verification.
+func WithTLSInsecure() Opt {
+	return func(c *Client) {
+		tlsConf := tls.Config{
+			InsecureSkipVerify: true, //nolint:gosec
+		}
+		if t, _ := c.httpClient.Transport.(*http.Transport); t == nil {
+			c.httpClient.Transport = &http.Transport{
+				TLSClientConfig: &tlsConf,
+			}
+		} else {
+			t.TLSClientConfig = &tlsConf
+		}
+	}
+}
+
+// WithUDS configures the client to connect over Unix Domain Socket.
+func WithUDS(socketPath string) Opt {
+	return func(c *Client) {
+		dialer := &net.Dialer{}
+		dialContext := func(ctx context.Context, _, _ string) (net.Conn, error) {
+			return dialer.DialContext(ctx, "unix", socketPath)
+		}
+		if t, _ := c.httpClient.Transport.(*http.Transport); t == nil {
+			c.httpClient.Transport = &http.Transport{
+				DialContext: dialContext,
+			}
+		} else {
+			t.DialContext = dialContext
+		}
+	}
+}
+
+func mergeWithReqOpts(ctx context.Context, evalContext *Context, opts *internal.ReqOpt) map[string]*structpb.Value {
+	evalContext.Data()
+	if opts == nil {
+		return evalContext.Data()
+	}
+	if evalContext == nil {
+		evalContext = NewContext().
+			WithIncludeMeta(opts.IncludeMeta)
+	}
+	if opts.AuxData != nil {
+		evalContext = evalContext.WithAuxData(opts.AuxData)
+	}
+	if opts.RequestIDGenerator != nil {
+		evalContext = evalContext.WithRequestID(opts.RequestIDGenerator(ctx))
+	}
+	return evalContext.Data()
+}
+
+func NewGRPCClient(address string, opts ...cerbos.Opt) (*Client, error) {
+	conf := internalgrpc.NewConfig(address)
+
+	for _, o := range opts {
+		o(conf)
+	}
+
+	grpcConn, err := internalgrpc.MkConn(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{stub: svcv1.NewAuthorizationServiceClient(grpcConn)}, nil
+}
+
+// NewClient creates a new AuthZEN HTTP client.
+// The baseURL should be the full URL to the Cerbos server (e.g., "https://pdp.example.com:3592").
+func NewClient(baseURL string, opts ...Opt) (*Client, error) {
```

Comment:
I'd swap this around so that `NewClient` creates a gRPC client by default.

BTW, if we use Connect, the configuration would be almost identical regardless of whether the user wants gRPC or HTTP. Unfortunately, it would be more work because we haven't implemented a Connect client yet. I'll create an issue for that.